File: js-algorithm-exercise/11_sort/quickSort.js
function quickSort(arr, left = 0, right = arr.length - 1) {
  // PIVOT 설정이 가능하게끔 DEFAULT VALUE
  function pivot(arr, start = 0, end = arr.length + 1) {
    // assigning pivot
    let pivot = arr[start];
    let swapIdx = start;
    console.log("THE PIVOT VALUE IS: ", pivot);
    //   where should the pivot be placed?
    //   in other words, how many values are smaller than pivot?
    for (let i = start + 1; i < arr.length; i++) {
      if (pivot > arr[i]) {
        console.log(arr);
        swapIdx++;
        [arr[i], arr[swapIdx]] = [arr[swapIdx], arr[i]];
      }
    }

    [arr[start], arr[swapIdx]] = [arr[swapIdx], arr[start]];

    return swapIdx;
  }
  // base case
  if (left < right) {
    let pivotIndex = pivot(arr, left, right);
    // left side
    quickSort(arr, left, pivotIndex - 1);
    // right side
    quickSort(arr, pivotIndex + 1, right);
  }

  return arr;
}

ex = [5, 2, 1, 8, 4, 7, 6, 3];
console.log(quickSort(ex));

File: js-algorithm-exercise/11_sort/quickSort.py
def quickSort(arr, start, end):
    start = 0
    end = len(arr)

    while start < end:
        pivotIdx = pivot(arr, start)

    # quickSort(arr, start, pivotIdx - 1)
    # quickSort(arr, pivotIdx + 1, end)

    return arr


def pivot(arr, start, end):
    pivotVal = arr[start]
    pivotIdx = start

    for i in range(1, end):
        print(arr)
        if pivotVal > arr[i]:
            pivotIdx += 1
            arr[i], arr[pivotIdx] = arr[pivotIdx], arr[i]
    arr[pivotIdx], arr[start] = arr[start], arr[pivotIdx]

    print(arr, pivotIdx)

    return pivotIdx


arr = [48, 0, 35, 32, 26, 27, 73]
quickSort(
    arr,
)
